From 4087a1fffe38106e10646606a27f10d40451862d Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Wed, 25 Jan 2017 16:24:52 +0100
Subject: [PATCH] dm rq: cope with DM device destruction while in
 dm_old_request_fn()
Git-commit: 4087a1fffe38106e10646606a27f10d40451862d
Patch-mainline: v4.10-rc8
References: FATE#321488

Fixes a crash in dm_table_find_target() due to a NULL struct dm_table
being passed from dm_old_request_fn() that races with DM device
destruction.

Reported-by: artem@flashgrid.io
Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Cc: stable@vger.kernel.org
Signed-off-by: Coly Li <colyli@suse.de>

---
 drivers/md/dm-rq.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/md/dm-rq.c b/drivers/md/dm-rq.c
index 9d7275f..6e702fc 100644
--- a/drivers/md/dm-rq.c
+++ b/drivers/md/dm-rq.c
@@ -779,6 +779,10 @@ static void dm_old_request_fn(struct request_queue *q)
 		int srcu_idx;
 		struct dm_table *map = dm_get_live_table(md, &srcu_idx);
 
+		if (unlikely(!map)) {
+			dm_put_live_table(md, srcu_idx);
+			return;
+		}
 		ti = dm_table_find_target(map, pos);
 		dm_put_live_table(md, srcu_idx);
 	}
-- 
2.10.2

